### ISSUE TITLE
What is the scope of ABI definitions?

### ISSUE THREAD

**Issue by pepyakin** [paritytech/ink#82]
I am creating this issue in order to not steer the discussion from the original issue. In https://github.com/paritytech/ink/issues/81 @Robbepop wrote. 

> In this folder we can place the (.json encoded) API description files of other contracts that we want to depend on with our contract. With this we maybe should consider renaming the term API description file since they are more than that by then.

When I read that I realized that we don't have a coherent story about ABI definitions, i.e. what ABI definitions are going to be used for. Do we want it to contain everything that produced by compilation process, e.g. #49, events descriptions (and in future other descriptions)? Or do we want to produce separate "ABI descriptions" for different use-cases (e.g. for block explorers we produce something that contains everything and for linking with others we can provide other). How does this play with the rest ecosystem? These ABI files are expected to be picked up by other languages?

**Comment by Robbepop** [comment]
Thanks for opening this issue!
This is important to talk about and even though I have already thought about it I couldn't come to a decent answer.
On the one hand we want a single file as output that just has all meta information about the compiled contract because this would be the simplest approach for users.
On the other hand side this could be a bad idea since then we may clutter lots of unneeded information everywhere. So maybe we do want to output different files with different purposes.

Let's discuss!

**Comment by xlc** [comment]
To me, the ABI file(s) should have two parts:
- The contract interface definition
  - type information
  - calls
  - events
  - storages
- The contract instantiation information
  - The contract code hash, address, network id
  
I think the contract interface definition should just be a single file. I don't see much downside on supply extra information (e.g. events and storages are not needed for dependent contract). The files are distributed off-chain and they are not large.

**Comment by Robbepop** [comment]
> - The contract interface definition
>
>   - type information
>   - calls
>   - events
>   - storages

This basically sums it up what is going to happen once https://github.com/paritytech/ink/pull/104 is merged.
